### ISSUE TITLE
Follow primary command not fixing postgresql.auto.conf

### ISSUE THREAD

**Issue by mirecekparecek** [pgpool/pgpool2#67]
Hello,

I use postgresql 15.6 and pgpool 4.5.2 after changin the home directory and setting up passwordless ssh with failover and followprimary it works well except for the postgresql.auto.conf it uses the wrong ip of the two node cluster when I use pcp_promote_node --switchover 

follow_primary_command = '/etc/pgpool-II/follow_primary.sh %d %h %p %D %m %H %M %P %r %R'

After successfully running the node is in waiting state unless I do show pool_nodes and after that it is in pgpool but the replication doesn't work because of the postgresql.auto.conf

Also I try to use pcp.conf but it works only for one user when i try to use pcp_* from different users it returns that the authentication failed even though there are multiple defined in the file pcp.conf and I defined all using this one-liner: echo 'user:'`pg_md5 password` >> /etc/pgpool-II/pcp.conf. How can it work for one user and not for others ?

Thank you very much for response

**Comment by pengbo0328** [comment]
The sample scripts included in the Pgpool-II package don't use postgresql.auto.conf. 
Below is an example:
https://www.pgpool.net/docs/latest/en/html/example-cluster.html

If you are using a custom script, you will need to modify it to configure the streaming replication settings.

**Comment by mirecekparecek** [comment]
I use replication on the postgres level - streaming replication and wanted to do switchover and it works but it misses the wal sender and receiver which is the only thing I have to change manually and restart the standby. But, the variables get set the primary ip address of the new primary correctly. Which is why I don't understand why it ends up like that.

**Comment by pengbo0328** [comment]
@mirecekparecek 

1. The sample script saves the settings for streaming replication in `${NODE_PGDATA}/myrecovery.conf`.
    It doesn't writes the settings for streaming replication to `postgresql.auto.conf`.

2. I think you used `pcp_promote_node --switchover -n <node_id>` to perform switchover.
   To use this command, you need to set `follow_primary_command`.

   https://www.pgpool.net/docs/latest/en/html/pcp-promote-node.html

**Comment by mirecekparecek** [comment]
Yes the follow_primary_command is set with the sample, the myrecovery.conf has set up primary_conninfo correctly but that's why I don't know why it puts wrong ip address to the postgresql.auto.conf and destroys the replication.

**Comment by pengbo0328** [comment]
> Yes the follow_primary_command is set with the sample, the myrecovery.conf has set up primary_conninfo correctly but that's why I don't know why it puts wrong ip address to the postgresql.auto.conf and destroys the replication.

Because the sample script doesn't use `pg_basebackup -R` option and doesn't use `ALTER SYSTEM` command, 
I think postgresql.auto.conf is modified by other software.

**Comment by mirecekparecek** [comment]
It's on test server where there is only a postgresql and pgpool for testing pgpool, nothing else.

**Comment by pengbo0328** [comment]
Pgpool-II doesn't use `postgresql.auto.conf` for streaming replication.
Did you configure streaming replication of PostgreSQL using `postgresql.auto.conf` by yourself during the initial setup?

**Comment by mirecekparecek** [comment]
I just changed it when I created the replication by pgbasebackup.Sent from HCL VerseBo Peng --- Re: [pgpool/pgpool2] Follow primary command not fixing postgresql.auto.conf (Issue #67) --- From:"Bo Peng" ***@***.***>To:"pgpool/pgpool2" ***@***.***>Cc:"mirecekparecek" ***@***.***>, "Mention" ***@***.***>Date:Thu, Aug 22, 2024 10:30 AMSubject:Re: [pgpool/pgpool2] Follow primary command not fixing postgresql.auto.conf (Issue #67)
Pgpool-II doesn't use postgresql.auto.conf for streaming replication.
Did you configure streaming replication of PostgreSQL using postgresql.auto.conf by yourself during the initial setup?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/pgpool/pgpool2/issues/67#issuecomment-2304083378",
"url": "https://github.com/pgpool/pgpool2/issues/67#issuecomment-2304083378",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by pengbo0328** [comment]
> I just changed it when I created the replication by pgbasebackup. 

I think you mean that you created standby using `pg_basebackup` with `-R` option. 
`-R` option will automatically create `primary_conninfo` in `postgresql.auto.conf`.

> Couldn't the issue lay in pg_rewind?

The follow_primary sample script uses pg_rewind. 
See more detail: https://git.postgresql.org/gitweb/?p=pgpool2.git;a=blob_plain;f=src/sample/scripts/follow_primary.sh.sample;hb=refs/heads/V4_5_STABLE

The problem is that the sample script doesn't use `ALTER SYSTEM` to change `primary_conninfo` in `postgresql.auto.conf`.
If you want to use `postgresql.auto.conf`, you can customize the sample script to use `ALTER SYSTEM` to change `primary_conninfo`.
 
> I first set up streaming replication in postgres then put pgpool above it. 

Alternatively, you need to configure primary_conninfo in the same way as shown in the sample script when you first set up the standby server.

```
cat > myrecovery.conf << EOT
primary_conninfo = 'host=${NEW_PRIMARY_NODE_HOST} port=${NEW_PRIMARY_NODE_PORT} user=${REPLUSER} application_name=${NODE_HOST} passfile=''/var/lib/pgsql/.pgpass'''
recovery_target_timeline = 'latest'
primary_slot_name = '${REPL_SLOT_NAME}'
EOT

vi postgresql.conf
(add the line below)
include_if_exists = 'myrecovery.conf'
```

**Comment by mirecekparecek** [comment]
Yes. you're right with everything, the thing is, both postgresql.auto.conf are right before i do switchover, but then when I do switchover the old primary has wrong ip in there, even though the variables that you can see in log for example %d, %D are set up right and I don't understand why after the switchover it ends up with wrong one because that makes it so that the new standby(old primary) can't connect, unless i manually change it.

**Comment by pengbo0328** [comment]
> but then when I do switchover the old primary has wrong ip in there, even though the variables that you can see in log for example %d, %D are set up right and I don't understand why after the switchover it ends up with wrong one because that makes it so that the new standby(old primary) can't connect, unless i manually change it.

Because pgpool doesn't change `postgresql.auto.conf`. The old settings in `postgresql.auto.conf` will override the settings used by pgpool (e.g. %d, %D).

I suggest configuring primary_conninfo in the same way as shown in the sample script when you first set up the standby server.

```
cat > myrecovery.conf << EOT
primary_conninfo = 'host=${NEW_PRIMARY_NODE_HOST} port=${NEW_PRIMARY_NODE_PORT} user=${REPLUSER} application_name=${NODE_HOST} passfile=''/var/lib/pgsql/.pgpass'''
recovery_target_timeline = 'latest'
primary_slot_name = '${REPL_SLOT_NAME}'
EOT

vi postgresql.conf
(add the line below)
include_if_exists = 'myrecovery.conf'
```

**Comment by jtorral** [comment]
I did this on my test system and it work

find ....

```
cat > ${RECOVERYCONF} << EOT
primary_conninfo = 'host=${NEW_PRIMARY_NODE_HOST} port=${NEW_PRIMARY_NODE_PORT} user=${REPLUSER} application_name=${NODE_HOST} passfile=''/var/lib/pgsql/.pgpass'''
recovery_target_timeline = 'latest'
primary_slot_name = '${REPL_SLOT_NAME}'
EOT
```


And add this  after the above

```
# Compensate for postgresql.auto.conf having old connection string by removing entries from postgresql.auto.conf

sed -i -e '/^primary_conninfo/d' -e '/^primary_slot_name/d' -e '/^recovery_target_timeline/d'  ${NODE_PGDATA}/postgresql.auto.conf
```
